### PR TITLE
Add get() and getHandle() to dependent ES records, and getTransientHandle() to EventSetup

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

This PR adds `get()` and `getHandle`() to dependent record enabling the same ESProduct retrieval pattern as from the non-dependent records, i.e. instead of
```cpp
edm::ESGetToken<BarData, BarRecord> barToken_;
...
std::unique_ptr<Foo> FooESProducer::produce(const FooRecord& iRecord) {
  const auto& barData = iRecord.getGetRecord<BarRecord>().get(barToken_);
  // or
  edm::ESHandle<BarData> barHandle = iRecord.getRecord<BarRecord>().get(barToken_);
  ...
}
```
one can do
```cpp
edm::ESGetToken<BarData, BarRecord> barToken_;
...
std::unique_ptr<FooData> FooESProducer::produce(const FooRecord& iRecord) {
  const auto& barData = iRecord.get(barToken_)
  // or
  edm::ESHandle<BarData> barHandle = iRecord.getHandle(barToken_);
  ...
}
```

This PR also adds `getTransientHandle()` to `EventSetup`, `EventSetupRecordImplementation`, and `DependentRecordImplementation` along the `getHandle()`. These functions are needed to support `edm::ESTransientHandle<T>` with `edm::ESGetToken`.

#### PR validation:

Framework compiles and relevant unit tests run.

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by Dr15Jones** [comment]
+1

File: FWCore/Framework/interface/DependentRecordImplementation.h [hunk 1]
```diff
@@ -76,6 +76,44 @@ class DependentRecordImplementation : public EventSetupRecordImplementation<Reco
...
       }
 
+      using EventSetupRecordImplementation<RecordT>::getHandle;
+
+      template<typename ProductT, typename DepRecordT>
+      ESHandle<ProductT> getHandle(ESGetToken<ProductT, DepRecordT> const& iToken) const {
+        //Make sure that DepRecordT is a type in ListT
+        using EndItrT = typename boost::mpl::end< ListT >::type;
+        using FoundItrT = typename boost::mpl::find< ListT, DepRecordT >::type;
+        static_assert(! std::is_same<FoundItrT, EndItrT>::value, "Trying to get a a product with an ESGetToken specifying a Record from another Record where the second Record is not dependent on the first Record.");
```
Comment (kpedro88) [inline FWCore/Framework/interface/DependentRecordImplementation.h]: fix duplicate "a" in this message and the two below ("Trying to get a a product")

Comment (makortel) [reply FWCore/Framework/interface/DependentRecordImplementation.h]: Good catch, thanks, fixed.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by Dr15Jones** [comment]
+1

**Comment by kpedro88** [comment]
+1

**Pull request merged**